### ISSUE TITLE
[Request] Automatically detect the best intensity and threads

### ISSUE THREAD

**Issue by qxxrbull** [fireice-uk/xmr-stak#1606]
When using AMD Vega 56 or 64
If using default intensity and threads setting , the performance is not the best.
(almost 1200~H/s per card)

The best parameter should be 2 threads 
and the intensity should be like 1600+2016、1792+1792 even 1932+1932

Is there any method or possibly to get the best amd.txt setting parameter that more friendly for beginner.

**Comment by Spudz76** [comment]
Auto config doesn't do multiple host threads, that would be a whole new addition.

**Comment by psychocrypt** [comment]
This can only be done by brute force testing all variants. It can also went
into the problem that rhe system crash e.g. if wr take all gpu memory and
the x server tries to allocate a little bit more.
We can put it on our long term agenda.

**Comment by Spudz76** [comment]
The only method I can imagine would be to run very short benchmarks to test and eliminate really slow choices quickly, and then retry the winners of that round with a longer benchmark so as to waste the least time drilling-down to the good neighborhood, and more time thoroughly testing within the zone.

More or less what I do by hand with `--benchmark` and why I added the `--benchwait` and `--benchwork` a little while ago, so as to run shorter homing tests (humans would be extinct by the time you tried "all" options at 30s+60s defaults).

Main issue is the shutdown/startup of a fresh backend each time.  There is no "live" tweaking on the kernel args, especially in the GPU backends which compile kernels based on the settings fed in as define literals at runtime.  Even doing my manual method with `--benchmark` sometimes sends a GPU into dreamworld, or CUDA's favorite locking it in P8 powersave mode, and I have to reboot to continue valid testing.  At least AMD stuff usually locks up harder and won't OpenCL again, when it doesn't like something?  Auto benchmark would churn along on that CUDA card that goes P8 on failure, because it still works fine, just half speed and it would disqualify a whole bunch of possibly good settings.  Unless it monitored the CUDA powerstates and etc which is even more stuff not here, to add.

cast-xmr must have a completely different core design where it can change tuning on the fly, or closer to on the fly than closing out CUDA/OpenCL/CPU and re-initializing from ground up each time.  Or perhaps a different tuning-kernel it runs which steps through and tests many settings in one load/run cycle, more of a simple benchmark testing the various op speeds and not the entirety of a hash-round.  From the op speeds you should be able to calculate a really close guess for optimal real-work hashing.  Pick the shortest sub-op methods and alignments and sizing and then put those together and it should be optimal.

**Comment by Simaex** [comment]
As long as we have dev fee I think that a community benchmark effort is possible. People that are willing to pay dev fee can have additional perk provided they permit to upload hardware configuration. It may be much better then multiple and lengthy test for every user.

**Comment by Spudz76** [comment]
Yes this feature would depend heavily on accurate reporting of cpu/gpu model detection info with metrics from config mapped to real results.  Also complicating any past device-vs-config-vs-rate measurements are all the new/changed algorithms, metrics should also be tracked per algo as they can/will differ.

Actually may be more sensible to enable by default if the donation is disabled, then the user donates the data metrics instead of hash work.  I disable my donations and instead come here and help out (annoy?)

Then since 2% is donated by default the data-metrics would be default off but able to be turned on (`--sendRates` or such).  Generally any 'phone-home' feature should be default off as best practice, as an opt-in - with the exception of the auto enable if donation is 0% which is a side case and needs also to be clearly defined in the readme/docs something like:
```
"xmr-stak will send non-personally-identifiable data regarding hardware classification data, configuration of those devices, and the resulting hash rate for that combination.  This data will be used for optimization purposes only, your IP address and any other identifiable data are not stored, only a fingerprint of the configuration for uniqueness keying."
```

Also hostname and IP would be fed into a one-way easy/small hashing function such as md5 or even crc32 as the fingerprint/uniqueness indicator to avoid much of the duplicated data (one entry per report reduced to more like one entry per rig, and without sending/storing any identifiable or potentially sensitive data).

**Comment by Simaex** [comment]
I beleive that such community effort may be of great use for upcoming GUI version. Having the ability to check you results against best known will help new users a lot.

**Comment by Spudz76** [comment]
I have three identical motherboard/cpu combos and they vary on the CPU rate total by 5H/s or so.
Identical!  I even adjusted them to make sure they were as identical as possible, and still a difference.

So I hope this doesn't just make users stop by and post 20 Issues about 5H/s """missing""" or whatever.  I can already hear myself typing "It's not missing, it's normal variance" 1000 times.

**Comment by Simaex** [comment]
Having multiple readings give you ability to calculate precision of the measurement. We do not need exact results, average and mean deviation is good enough. Another feature - result rating, ex "you results are better than 95% of users of current version". It is not about programming, it is marketing|UI to get it right. I surely agree that it will not stop users from posting strange questions. But surely it will help people that are willing to do some tuning.

**Comment by hotaru2k3** [comment]
I'd be happy if the auto config would just stop always coming up with settings for Nvidia GPUs that make the miner crash with a CUDA error within about 30 seconds. it seems like the auto config settings are always too low on AMD GPUs and too high on Nvidia.

**Comment by qxxrbull** [comment]
I think we can get the parameter form xmrstak.com ?
